Block Library: Add new Comment Reply Link block

DAreRodz opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This PR implements a block which adds a link to reply to a specific comment. The implementation will follow the design explained in the linked issue.

Will close #30576.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

<details>
<summary><b>Typography, align and Color settings</b></summary>
</br>

1. Add a post comment block (using an existing comment ID).
2. Add an inner post comment reply link block.
3. Confirm that the color and typography options work.
4. Confirm that the text align option works.

**NOTE:** there is an issue related to global styles, check https://github.com/WordPress/gutenberg/pull/35774#issuecomment-952046182

</details>

<details>
<summary><b>Threads depth limit</b></summary>
</br>

1. Create a comment (comment 1).
2. Create another comment, replying the previous one (comment 2).
3. Go to site editor and, in the Single Post template, add a post comment block with the reply link for each comment.
4. In the frontend, go to a post and confirm that both render the correct reply link.
5. Go to Discussions Settings and change the "levels deep" setting to `2`.
6. Confirm that only comment 1 renders a reply link.
7. Go back to Discussions Settings and disable threaded comments entirely.
8. Confirm that none of the comments render a reply link.

</details>

<details>
<summary><b>Invalid or missing comment ID</b></summary>
</br>

1. Add a comment block and set an invalid comment ID.
2. In the frontend, confirm that the block is not rendered.

</details>

<details>
<summary><b>Only registered and logged in users can comment</b></summary>
</br>

1. In Discussions Settings, enable "Users must be registered and logged in to comment"
2. Add a comment block in the Single Post template, and set a valid comment ID.
3. Visit a post in incognito.
4. confirm that the rendered reply link shows a `Log in to reply` message or similar.

</details>

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

DAreRodz committed:
Comment Reply: Init block files

DAreRodz committed:
Remove unexpected throw

DAreRodz committed:
Comment Reply: Render a clickable link

- It uses `get_comment_reply_link` to generate the link.
- That function already cover most of the cases, except comments depth.

DAreRodz committed:
Comment Reply: Add style settings

DAreRodz committed:
Comment Reply: Create block fixture files

DAreRodz committed:
Comment Reply: Fix phpcs errors

DAreRodz committed:
Comment Reply: Fix some comments

DAreRodz commented:
Hey! I want to explain what I did so far. It would be great to receive some feedback here. :blush:

To generate the reply link, I'm using the `get_comment_reply_link()` function that exists on WordPress. That function already covers by default some of the cases [mentioned](https://github.com/WordPress/gutenberg/issues/30576#issuecomment-923863803) by @SantosGuillamot in the [issue](https://github.com/WordPress/gutenberg/issues/30576#issuecomment-923863803), like changing the text when a user is not logged in (if only users can comment), or hiding the link when comments are disabled.

To hide the reply link when the comment has a certain depth, I've created a loop that iterates over all the comment's ancestors. Is this loop OK, or is there a better way to know the depth value of a comment?

https://github.com/WordPress/gutenberg/blob/55594dac09ba91b36fbda32fbe70bf0b3a9d84cb/packages/block-library/src/post-comment-reply-link/index.php#L26-L36

And also, related to that: should the reply link be hidden in the editor as well when reached a certain depth? In that case, I guess the same loop should be replicated in the `edit.js` file to compute the depth value.

cc: @gziolo @ntsekouras

DAreRodz commented on packages/icons/src/library/post-comment-reply-link.js:
```diff
@@ -0,0 +1,28 @@
+		width="24"
+		height="24"
+		viewBox="0 0 24 24"
+		fill="none"
```
I've just copied the icon from the design, and seems that this `fill` attribute is overwritten by
```css
.block-editor-block-icon.has-colors svg {
    fill: currentColor;
}
```

making the icon look like this:

![image](https://user-images.githubusercontent.com/6917969/138759693-29b7f2f9-e0cb-4102-ad5b-69ca7b86d238.png)

@jameskoster, do you know if there is a quick way to fix this or does the svg need to be changed?

jameskoster replied:
That looks like the wrong svg code anyway, let me grab the correct one:

```
<svg width="24" height="24" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M6.68822 10.625L6.24878 11.0649L5.5 11.8145L5.5 5.5L12.5 5.5V8L14 6.5V5C14 4.44772 13.5523 4 13 4H5C4.44772 4 4 4.44771 4 5V13.5247C4 13.8173 4.16123 14.086 4.41935 14.2237C4.72711 14.3878 5.10601 14.3313 5.35252 14.0845L7.31 12.125H8.375L9.875 10.625H7.31H6.68822ZM14.5605 10.4983L11.6701 13.75H16.9975C17.9963 13.75 18.7796 14.1104 19.3553 14.7048C19.9095 15.2771 20.2299 16.0224 20.4224 16.7443C20.7645 18.0276 20.7543 19.4618 20.7487 20.2544C20.7481 20.345 20.7475 20.4272 20.7475 20.4999L19.2475 20.5001C19.2475 20.4191 19.248 20.3319 19.2484 20.2394V20.2394C19.2526 19.4274 19.259 18.2035 18.973 17.1307C18.8156 16.5401 18.586 16.0666 18.2778 15.7483C17.9909 15.4521 17.5991 15.25 16.9975 15.25H11.8106L14.5303 17.9697L13.4696 19.0303L8.96956 14.5303L13.4394 9.50171L14.5605 10.4983Z" />
</svg>
```

I assume we'll add this to the icon library as a follow up?

DAreRodz replied:
I've updated the icon with the code you shared. Now it works.

![image](https://user-images.githubusercontent.com/6917969/138947856-f3e3081b-29f8-4eb6-ae1a-cb436d8267fc.png)

> I assume we'll add this to the icon library as a follow up?

Already added to the icon library (not sure if that's what you mean) :smile:

gziolo commented on packages/block-library/src/post-comment-reply-link/edit.js:
```diff
@@ -0,0 +1,59 @@
+			{ blockControls }
+			<div { ...blockProps }>
+				<a
+					href="#comments"
```
Let's align with other blocks:
```suggestion
					href="#comment-reply-pseudo-link"
```

gziolo commented on packages/block-library/src/post-comment-reply-link/index.php:
```diff
@@ -0,0 +1,72 @@
+	$parent_id = $comment->comment_parent;
+
+	// Compute comment's depth iterating over its ancestors.
+	while ( 0 !== $parent_id ) {
```
@aristath, can you think of a better way to compute the depth of the current comment?

The only concern I have here is that I would feel better if the check for the `$parent_id` wasn't so strict about it being a number. I would feel more comfortable seeing the loop terminated as soon as `$parent_id` has falsy value.
```suggestion
	while ( ! empty( $parent_id ) ) {
```

gziolo replied:
I run into the issue like that:

<img width="1754" alt="Screen Shot 2021-10-26 at 17 01 26" src="https://user-images.githubusercontent.com/699132/138906456-eb1f5b18-6f69-49cf-b640-67074ea84125.png">

gziolo replied:
It happens in two cases, when there is a comment selected with an invalid id, or when a comment doesn't have any parents.

gziolo commented on packages/block-library/src/post-comment-reply-link/index.php:
```diff
@@ -0,0 +1,72 @@
+		return '';
+	}
+
+	$comment = get_comment( $block->context['commentId'] );
```
When the comment id is invalid, then `$comment` is not an object so it will throw a notice in line 30:

<img width="1766" alt="Screen Shot 2021-10-26 at 17 08 17" src="https://user-images.githubusercontent.com/699132/138907680-69bcc632-3c7b-44ea-81d7-130b7d2dd5ab.png">

We can account for that similar to other blocks with:
```suggestion
	$comment = get_comment( $block->context['commentId'] );
	if ( empty( $comment ) {
		return '';
	}
```

gziolo commented:
I found two issues related to (Global) Styles that don't seem related to this PR.

Background color + link color trigger contrast checker warning when they shouldn't:

<img width="279" alt="Screen Shot 2021-10-26 at 16 58 49" src="https://user-images.githubusercontent.com/699132/138908973-7b03c82b-040b-4050-aba6-d8a5d380c5a2.png">

I think the checker compares the background color selected and the default text color read from the same `div`, but we don't care about it at all.

The selected font family has class names applied correctly, but it isn't reflected in the computer font family in the editor:
<img width="1151" alt="Screen Shot 2021-10-26 at 17 14 43" src="https://user-images.githubusercontent.com/699132/138909159-033b8aa7-5cfa-43ed-8596-0a8b3b7c1595.png">

It seems to be correct on the frontend though:

<img width="472" alt="Screen Shot 2021-10-26 at 17 18 27" src="https://user-images.githubusercontent.com/699132/138909495-8c4be1c6-d180-47ab-b14d-212db928d325.png">

From my perspective, the only blockers are those related to PHP notices and most importantly the one that produces the fatal error.

DAreRodz committed:
Icons: Fix comment reply link

DAreRodz committed:
Comment Reply: Change href in editor

Using `#comment-reply-pseudo-link` for consistency.
That pattern is being used in other blocks, e.g.
Post Comment Author or Post Comment Date

DAreRodz committed:
Comment Reply: Fix SSR for some cases

The cases are:
- The comment ID passed is invalid
- The comment doesn't have any parents
- There is no link to render

DAreRodz commented:
Thanks for your review, @gziolo. I've fixed all the issues you found. (*)
BTW, I was wondering if it is worth adding PHP tests for the SSR of this block ― something similar to [/phpunit/class-block-library-navigation-link-test.php](https://github.com/WordPress/gutenberg/blob/cd0003c0f900bf29b0c9694f39a894318b49a9d4/phpunit/class-block-library-navigation-link-test.php) :thinking:

(*) EDIT: all except those related to global styles.

gziolo reviewed (approved):
Nice work on this PR 👍🏻 

It still needs a rebase after I landed the Comment Author Avatar block that changes whitespaces in the `lib/blocks.php` file.

> BTW, I was wondering if it is worth adding PHP tests for the SSR of this block ― something similar to /phpunit/class-block-library-navigation-link-test.php 🤔

The Navigation block is far more complex. We could add PHP tests for the Comments Query Loop block once it is fully functional that tests the `render_callback` behavior. Although, I feel like e2e tests are simpler to write and give more confidence because they ensure that the user experience is also good. We can look at that after we have all the required blocks in place.

gziolo commented on packages/block-library/src/post-comment-reply-link/index.php:
```diff
@@ -0,0 +1,80 @@
+
+	$classes = '';
+	if ( isset( $attributes['textAlign'] ) ) {
+		$classes .= 'has-text-align-' . $attributes['textAlign'];
```
We should also escape the text align value in case it gets updated manually in the post content.

gziolo replied:
You can ignore my previous comment. It's sanitized inside `get_block_wrapper_attributes` function 👍🏻

gziolo committed:
Merge branch 'trunk' into add/comment-reply-link

gziolo commented:
I tested again the issues raised earlier and it looks like the font family handling is no longer an issue:

> The selected font family has class names applied correctly, but it isn't reflected in the computer font family in the editor:

This is how it looks like now:

<img width="1183" alt="Screenshot 2021-12-21 at 09 58 16" src="https://user-images.githubusercontent.com/699132/146901630-422f07e6-f26d-4907-b247-84c66b5fd28a.png">

The issue with the color contrast is still present and I'm about to file a new issue for that to discuss possible solutions.
